dask/array/core.py

```diff
@@ -2062,7 +2062,7 @@ def asarray(array):
     """
     if not isinstance(array, Array):
         name = 'asarray-' + tokenize(array)
-        if not hasattr(array, 'shape'):
+        if isinstance(getattr(array, 'shape', None), Iterable):
```

Comment:
To be clear you're talking about the change of the predicate from "hasattr shape" to "shape is an iterable"?  If so then can you describe more about what an h5py.Dataset shape looks like?

Do we not test that h5py.Datasets stay h5py.Datasets until necessary?